dask_match/repartition.py

```diff
@@ -0,0 +1,332 @@
+import functools
+from operator import getitem
+from pprint import pformat
+
+import pandas as pd
+import numpy as np
+from pandas.api.types import is_datetime64_any_dtype, is_numeric_dtype
+from tlz import unique
+
+from dask.dataframe import methods
+from dask.dataframe.core import split_evenly
+from dask.dataframe.utils import is_series_like
+from matchpy import CustomConstraint, Pattern, ReplacementRule, Wildcard
+
+from dask_match.expr import Expr, Filter, Projection
+
+
+class Repartition(Expr):
+    """Abstract repartitioning expression"""
+
+    _parameters = ["frame", "n", "new_divisions", "force"]
+
+    @property
+    def _meta(self):
+        return self.frame._meta
+
+    def _divisions(self):
+        if self.n is not None:
+            return self.simplify()._divisions()
+        return self.new_divisions
+
+    def simplify(self):
+        if self.n is not None:
+            if self.n < self.frame.npartitions:
+                return ReducePartitionCount(self.frame, self.n)
```

Comment:
I suggest that all of these classes should maybe start with `Repartition`, like `RepartitionToMorePartitions` and `RepartitionToFewerPartitions`.  Mostly I care about Repartition being in the name, and probably the first term.

Reply:
Cool - Make sense to me.